File: backend/src/main/java/com/horain/chat/SseEmitterStreamEventWriter.java
package com.horain.chat;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.http.MediaType;
import org.springframework.web.servlet.mvc.method.annotation.SseEmitter;

import java.io.IOException;
import java.util.List;
import java.util.Map;
import java.util.UUID;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

/**
 * StreamEventWriter implementation that writes SSE events to an SseEmitter.
 * Event format: "chunk" (data = JSON {"text": "delta"}), "done" (data = JSON payload), "error" (data = JSON {"message": "..."}).
 */
public class SseEmitterStreamEventWriter implements StreamEventWriter {

    private static final Logger log = LoggerFactory.getLogger(SseEmitterStreamEventWriter.class);
    private static final long EMITTER_TIMEOUT_MS = 300_000L;

    private final SseEmitter emitter;
    private final ObjectMapper objectMapper;
    private boolean completed;

    public SseEmitterStreamEventWriter(SseEmitter emitter, ObjectMapper objectMapper) {
        this.emitter = emitter;
        this.emitter.onTimeout(() -> {
            if (!completed) {
                completed = true;
                log.debug("SSE emitter timed out");
            }
        });
        this.emitter.onCompletion(() -> completed = true);
        this.objectMapper = objectMapper;
        this.completed = false;
    }

    @Override
    public void sendChunk(String text) {
        if (completed) return;
        try {
            String data = objectMapper.writeValueAsString(Map.of("text", text != null ? text : ""));
            emitter.send(SseEmitter.event().name("chunk").data(data, MediaType.APPLICATION_JSON));
        } catch (IOException e) {
            log.warn("Failed to send chunk: {}", e.getMessage());
            completeWithError(e.getMessage());
        }
    }

    @Override
    public void sendDone(String assistantMessage, List<ToolCallRecord> toolCalls, List<Integer> toolCallIterations,
                         Object data, UUID turnId, String reasoningText, Long reasoningDurationMs) {
        if (completed) return;
        try {
            List<Map<String, Object>> toolCallsDto = toolCalls != null
                    ? IntStream.range(0, toolCalls.size())
                    .mapToObj(i -> {
                        ToolCallRecord tc = toolCalls.get(i);
                        Map<String, Object> m = new java.util.HashMap<>(Map.of(
                                "name", tc.name() != null ? tc.name() : "",
                                "arguments", tc.arguments() != null ? tc.arguments() : "",
                                "result", tc.result() != null ? tc.result() : ""));
                        if (toolCallIterations != null && i < toolCallIterations.size()) {
                            m.put("iterationIndex", toolCallIterations.get(i));
                        }
                        return m;
                    })
                    .collect(Collectors.toList())
                    : List.of();
            Map<String, Object> payload = new java.util.HashMap<>(Map.of(
                    "assistantMessage", assistantMessage != null ? assistantMessage : "",
                    "toolCalls", toolCallsDto,
                    "data", data != null ? data : Map.of()));
            if (turnId != null) {
                payload.put("turnId", turnId.toString());
            }
            if (reasoningText != null && !reasoningText.isBlank()) {
                payload.put("reasoningText", reasoningText);
            }
            if (reasoningDurationMs != null && reasoningDurationMs >= 0) {
                payload.put("reasoningDurationMs", reasoningDurationMs);
            }
            String json = objectMapper.writeValueAsString(payload);
            emitter.send(SseEmitter.event().name("done").data(json, MediaType.APPLICATION_JSON));
            emitter.complete();
            completed = true;
        } catch (JsonProcessingException e) {
            log.warn("Failed to serialize done payload: {}", e.getMessage());
            completeWithError(e.getMessage());
        } catch (IOException e) {
            log.warn("Failed to send done: {}", e.getMessage());
            completeWithError(e.getMessage());
        }
    }

    @Override
    public void sendReasoningChunk(String text) {
        if (completed) return;
        try {
            String data = objectMapper.writeValueAsString(Map.of("text", text != null ? text : ""));
            emitter.send(SseEmitter.event().name("reasoning_chunk").data(data, MediaType.APPLICATION_JSON));
        } catch (IOException e) {
            log.warn("Failed to send reasoning_chunk: {}", e.getMessage());
            completeWithError(e.getMessage());
        }
    }

    @Override
    public void sendError(String message) {
        completeWithError(message);
    }

    @Override
    public void sendToolCall(ToolCallRecord record, int iterationIndex) {
        if (completed || record == null) return;
        try {
            Map<String, Object> payload = new java.util.HashMap<>(Map.of(
                    "name", record.name() != null ? record.name() : "",
                    "arguments", record.arguments() != null ? record.arguments() : "",
                    "result", record.result() != null ? record.result() : "",
                    "iterationIndex", iterationIndex));
            String json = objectMapper.writeValueAsString(payload);
            emitter.send(SseEmitter.event().name("tool_call").data(json, MediaType.APPLICATION_JSON));
        } catch (JsonProcessingException e) {
            log.warn("Failed to serialize tool_call payload: {}", e.getMessage());
        } catch (IOException e) {
            log.warn("Failed to send tool_call: {}", e.getMessage());
        }
    }

    @Override
    public void sendAssistantSegment(String text, int iterationIndex) {
        if (completed) return;
        try {
            Map<String, Object> payload = new java.util.HashMap<>(Map.of(
                    "text", text != null ? text : "",
                    "iterationIndex", iterationIndex));
            String json = objectMapper.writeValueAsString(payload);
            emitter.send(SseEmitter.event().name("assistant_segment").data(json, MediaType.APPLICATION_JSON));
        } catch (IOException e) {
            log.warn("Failed to send assistant_segment: {}", e.getMessage());
            completeWithError(e.getMessage());
        }
    }

    private void completeWithError(String message) {
        if (completed) return;
        try {
            String data = objectMapper.writeValueAsString(Map.of("message", message != null ? message : "Unknown error"));
            emitter.send(SseEmitter.event().name("error").data(data, MediaType.APPLICATION_JSON));
        } catch (Exception e) {
            log.warn("Failed to send error event: {}", e.getMessage());
        }
        try {
            emitter.completeWithError(new RuntimeException(message));
        } catch (Exception e) {
            log.debug("Emitter already completed: {}", e.getMessage());
        }
        completed = true;
    }
}

File: backend/src/main/kotlin/com/horain/chat/SseEmitterStreamEventWriter.kt
package com.horain.chat

import com.fasterxml.jackson.core.JsonProcessingException
import com.fasterxml.jackson.databind.ObjectMapper
import org.slf4j.LoggerFactory
import org.springframework.http.MediaType
import org.springframework.web.servlet.mvc.method.annotation.SseEmitter
import java.io.IOException
import java.util.UUID

/**
 * StreamEventWriter implementation that writes SSE events to an SseEmitter.
 * Event format: "chunk" (data = JSON {"text": "delta"}), "done" (data = JSON payload), "error" (data = JSON {"message": "..."}).
 */
class SseEmitterStreamEventWriter(
    private val emitter: SseEmitter,
    private val objectMapper: ObjectMapper
) : StreamEventWriter {

    private var completed: Boolean = false

    init {
        emitter.onTimeout {
            if (!completed) {
                completed = true
                log.debug("SSE emitter timed out")
            }
        }
        emitter.onCompletion { completed = true }
    }

    override fun sendChunk(text: String) {
        if (completed) return
        try {
            val data = objectMapper.writeValueAsString(mapOf("text" to (text.ifBlank { "" })))
            emitter.send(SseEmitter.event().name("chunk").data(data, MediaType.APPLICATION_JSON))
        } catch (e: IOException) {
            log.warn("Failed to send chunk: {}", e.message)
            completeWithError(e.message)
        }
    }

    override fun sendDone(
        assistantMessage: String,
        toolCalls: List<ToolCallRecord>,
        toolCallIterations: List<Int>?,
        data: Any?,
        turnId: UUID?,
        reasoningText: String?,
        reasoningDurationMs: Long?,
        modelName: String?,
        conversationId: UUID
    ) {
        if (completed) return
        try {
            val toolCallsDto =
                toolCalls.mapIndexed { i, tc ->
                    val m = hashMapOf<String, Any>(
                        "name" to tc.name,
                        "arguments" to tc.arguments,
                        "result" to tc.result
                    )
                    if (toolCallIterations != null && i < toolCallIterations.size) {
                        m["iterationIndex"] = toolCallIterations[i]
                    }
                    m
                }
            val payload = hashMapOf<String, Any>(
                "assistantMessage" to (assistantMessage.ifBlank { "" }),
                "toolCalls" to toolCallsDto,
                "data" to (data ?: emptyMap<String, Any>())
            )
            if (turnId != null) {
                payload["turnId"] = turnId.toString()
            }
            if (!reasoningText.isNullOrBlank()) {
                payload["reasoningText"] = reasoningText
            }
            if (reasoningDurationMs != null && reasoningDurationMs >= 0) {
                payload["reasoningDurationMs"] = reasoningDurationMs
            }
            if (!modelName.isNullOrBlank()) {
                payload["modelName"] = modelName
            }
            payload["conversationId"] = conversationId.toString()
            val json = objectMapper.writeValueAsString(payload)
            emitter.send(SseEmitter.event().name("done").data(json, MediaType.APPLICATION_JSON))
            emitter.complete()
            completed = true
        } catch (e: JsonProcessingException) {
            log.warn("Failed to serialize done payload: {}", e.message)
            completeWithError(e.message)
        } catch (e: IOException) {
            log.warn("Failed to send done: {}", e.message)
            completeWithError(e.message)
        }
    }

    override fun sendReasoningChunk(text: String) {
        if (completed) return
        try {
            val data = objectMapper.writeValueAsString(mapOf("text" to text.ifBlank { "" }))
            emitter.send(SseEmitter.event().name("reasoning_chunk").data(data, MediaType.APPLICATION_JSON))
        } catch (e: IOException) {
            log.warn("Failed to send reasoning_chunk: {}", e.message)
            completeWithError(e.message)
        }
    }

    override fun sendReasoningPhaseDone(reasoningDurationMs: Long?) {
        if (completed) return
        try {
            val data = hashMapOf<String, Any>()
            if (reasoningDurationMs != null && reasoningDurationMs >= 0) {
                data["reasoningDurationMs"] = reasoningDurationMs
            }
            val json = objectMapper.writeValueAsString(data)
            emitter.send(SseEmitter.event().name("reasoning_phase_done").data(json, MediaType.APPLICATION_JSON))
        } catch (e: IOException) {
            log.warn("Failed to send reasoning_phase_done: {}", e.message)
            completeWithError(e.message)
        }
    }

    override fun sendModelName(modelName: String) {
        if (completed || modelName.isBlank()) return
        try {
            val data = objectMapper.writeValueAsString(mapOf("model" to modelName))
            emitter.send(SseEmitter.event().name("model").data(data, MediaType.APPLICATION_JSON))
        } catch (e: IOException) {
            log.warn("Failed to send model event: {}", e.message)
        }
    }

    override fun sendError(message: String) {
        completeWithError(message)
    }

    override fun sendToolCall(record: ToolCallRecord, iterationIndex: Int) {
        if (completed) return
        try {
            val payload = mapOf(
                "name" to record.name,
                "arguments" to record.arguments,
                "result" to record.result,
                "iterationIndex" to iterationIndex
            )
            val json = objectMapper.writeValueAsString(payload)
            emitter.send(SseEmitter.event().name("tool_call").data(json, MediaType.APPLICATION_JSON))
        } catch (e: JsonProcessingException) {
            log.warn("Failed to serialize tool_call payload: {}", e.message)
        } catch (e: IOException) {
            log.warn("Failed to send tool_call: {}", e.message)
        }
    }

    override fun sendAssistantSegment(text: String, iterationIndex: Int) {
        if (completed) return
        try {
            val payload = mapOf(
                "text" to text.ifBlank { "" },
                "iterationIndex" to iterationIndex
            )
            val json = objectMapper.writeValueAsString(payload)
            emitter.send(SseEmitter.event().name("assistant_segment").data(json, MediaType.APPLICATION_JSON))
        } catch (e: IOException) {
            log.warn("Failed to send assistant_segment: {}", e.message)
            completeWithError(e.message)
        }
    }

    private fun completeWithError(message: String?) {
        if (completed) return
        try {
            val data = objectMapper.writeValueAsString(
                mapOf("message" to (message?.ifBlank { null } ?: "Unknown error"))
            )
            emitter.send(SseEmitter.event().name("error").data(data, MediaType.APPLICATION_JSON))
        } catch (e: Exception) {
            log.warn("Failed to send error event: {}", e.message)
        }
        try {
            emitter.completeWithError(RuntimeException(message ?: "Unknown error"))
        } catch (e: Exception) {
            log.debug("Emitter already completed: {}", e.message)
        }
        completed = true
    }

    companion object {
        private val log = LoggerFactory.getLogger(SseEmitterStreamEventWriter::class.java)
    }
}
